File: macos/Sources/SnapBriefCore/Editing/SessionHistory.swift
import Foundation

/// Port of `src/SnapBrief.Core/Editing/SessionHistory.cs`.
///
/// C# uses `Stack<T>` (`_undo`/`_redo`) with `Push`/`Pop`; Swift arrays used as a LIFO stack via
/// `append`/`removeLast` are the direct equivalent. Record equality (`Current == next`) in C#
/// compares all properties structurally except that `ImmutableArray<T>.Equals` is reference-based
/// for the underlying array; `SnapBriefSession`'s synthesized `Equatable` here is fully
/// structural, which is a strictly finer-grained (never coarser) comparison and produces the same
/// observable skip-if-unchanged behavior for every call site in this codebase.
public final class SessionHistory {
    private var undoStack: [SnapBriefSession] = []
    private var redoStack: [SnapBriefSession] = []
    private let timeProvider: TimeProvider

    public private(set) var current: SnapBriefSession

    public init(initial: SnapBriefSession, timeProvider: TimeProvider = SystemTimeProvider()) {
        self.current = initial
        self.timeProvider = timeProvider
    }

    public var canUndo: Bool { !undoStack.isEmpty }
    public var canRedo: Bool { !redoStack.isEmpty }

    /// Port of `Apply(Func<SnapBriefSession, SnapBriefSession> operation)`.
    public func apply(_ operation: (SnapBriefSession) throws -> SnapBriefSession) rethrows {
        let next = try operation(current)
        if next == current {
            return
        }

        undoStack.append(current)
        redoStack.removeAll()
        current = next
    }

    @discardableResult
    public func undo() -> Bool {
        guard canUndo else { return false }

        let previous = undoStack.removeLast()
        redoStack.append(current)
        current = restoreAsNewRevision(previous, current)
        return true
    }

    @discardableResult
    public func redo() -> Bool {
        guard canRedo else { return false }

        let next = redoStack.removeLast()
        undoStack.append(current)
        current = restoreAsNewRevision(next, current)
        return true
    }

    private func restoreAsNewRevision(_ snapshot: SnapBriefSession, _ current: SnapBriefSession) -> SnapBriefSession {
        var restored = snapshot
        restored.revision = current.revision + 1
        restored.modifiedAtUtc = timeProvider.utcNow()
        return restored
    }
}

File: src/SnapBrief.Core/Editing/SessionHistory.cs
using SnapBrief.Core.Models;

namespace SnapBrief.Core.Editing;

public sealed class SessionHistory
{
    private readonly Stack<SnapBriefSession> _undo = new();
    private readonly Stack<SnapBriefSession> _redo = new();
    private readonly TimeProvider _timeProvider;

    public SessionHistory(SnapBriefSession initial, TimeProvider? timeProvider = null)
    {
        Current = initial;
        _timeProvider = timeProvider ?? TimeProvider.System;
    }

    public SnapBriefSession Current { get; private set; }
    public bool CanUndo => _undo.Count > 0;
    public bool CanRedo => _redo.Count > 0;

    public void Apply(Func<SnapBriefSession, SnapBriefSession> operation)
    {
        ArgumentNullException.ThrowIfNull(operation);
        var next = operation(Current);
        if (ReferenceEquals(Current, next) || Current == next)
        {
            return;
        }

        _undo.Push(Current);
        _redo.Clear();
        Current = next;
    }

    public bool Undo()
    {
        if (!CanUndo)
        {
            return false;
        }

        var previous = _undo.Pop();
        _redo.Push(Current);
        Current = RestoreAsNewRevision(previous, Current);
        return true;
    }

    public bool Redo()
    {
        if (!CanRedo)
        {
            return false;
        }

        var next = _redo.Pop();
        _undo.Push(Current);
        Current = RestoreAsNewRevision(next, Current);
        return true;
    }

    private SnapBriefSession RestoreAsNewRevision(SnapBriefSession snapshot, SnapBriefSession current) =>
        snapshot with
        {
            Revision = checked(current.Revision + 1),
            ModifiedAtUtc = _timeProvider.GetUtcNow()
        };
}
